### DISCUSSION TITLE
Memory spike while coalesce and write operation happens on multi frame GIF

### DISCUSSION THREAD

**Discussion by shivabhadauria** [dlemstra/Magick.NET#1039] [Help]
Hi @dlemstra & Team,

I have been trying to resize my multi-frame (415) gif and my code runs in a docker environment with a Memory restriction of 512 MB.
However, I noticed that the memory spikes to 920 MB+ while using Magick.NET-Q16-AnyCPU library to resize to 100 * 100.
Based on some articles that I came across, I changed it to Magick.NET-Q8-AnyCPU and it spikes to 520 MB+.
This works in a normal console application but when I include it in my container-based service, it crashes.
Also, if I resize it to say 400 * 400, it increases the size of the saved file which is not desirable. 
Attached is the reference gif that I used.
![1dpmw](https://user-images.githubusercontent.com/32168989/135261220-4954390d-20ba-43a2-89f1-2848351b84c9.gif)

The calculation segment I had to refer from here- https://stackoverflow.com/questions/10337800/resize-with-crop-using-imagemagick-net-and-c-sharp
I am struggling to optimize this to consume as little memory as possible together with a fast response time else it gives timeout.

**Code snippet:**
```
            var newWidth = 200;
            var newHeight = 200;
            Stopwatch watch = new Stopwatch();
            watch.Start();
            using (var collection = new MagickImageCollection(new FileInfo("filepath")))
            {
                collection.Coalesce();
                MagickGeometry intermediate_geo = new();
                MagickGeometry final_geo = new();
                var refImage = collection[0];
                if (refImage.Height != newHeight || refImage.Width != newWidth)
                {
                    decimal result_ratio = (decimal)newHeight / (decimal)newWidth;
                    decimal current_ratio = (decimal)refImage.Height / (decimal)refImage.Width;
                    bool preserve_width = false;
                    if (current_ratio > result_ratio)
                    {
                        preserve_width = true;
                    }

                    int new_width;
                    int new_height;
                    if (preserve_width)
                    {
                        new_width = newWidth;
                        new_height = (int)Math.Round((decimal)(current_ratio * new_width));
                    }
                    else
                    {
                        new_height = newHeight;
                        new_width = (int)Math.Round((decimal)(new_height / current_ratio));
                    }

                    String geomStr = newWidth.ToString() + "x" + newHeight.ToString();
                    String newGeomStr = new_width.ToString() + "x" + new_height.ToString();
                    intermediate_geo = new MagickGeometry(newGeomStr);
                    final_geo = new MagickGeometry(geomStr);
                }

                foreach (var image in collection)
                {
                        image.Scale(intermediate_geo);
                        image.Resize(intermediate_geo);
                        image.Crop(final_geo, Gravity.Center);
                        image.RePage();
                }
                collection.Write("filepath");
            }
            watch.Stop();
            Console.WriteLine("Elapsed Time for approach 2 is {0} ms", watch.ElapsedMilliseconds);
```

**Comment by dlemstra** [comment]
The large memory footprint is happening because of the way the `collection.Coalesce()` method works. This will create a new sequeunce of images and replaces the images in the collection. But this does mean that at some point in time you will have both image sequences in memory.

A possible solution would be to only resize and crop the images and adjust the `Page` attributes of the image to fix the offsets.

p.s. I don't understand why you do both a `Scale` and a `Resize`.

**Reply by shivabhadauria** [reply]
@dlemstra thank you very much for your response.
**Scale** is definitely a redundant step and I added that by mistake just to check the change in file size saved but I removed that eventually.
From the above snippet, I just **removed the coalesce step**, and the rest code was the same.
Memory for 100 * 100 fell to almost 300 MB.
Since I just need to write this file to a location for my utility, is whatever is listed sufficient or eventually, some dispose for the collection or images needs to be called?
**Using** statement shall automatically dispose of the collection, but is there a way to reduce it even more in this scenario?
```
using (var collection = new MagickImageCollection(new FileInfo("filepath")))
            {
                // Code for intermediate_geo & final_geo calculations
```
                foreach (var image in collection)
                {
                        image.Resize(intermediate_geo);
                        image.Crop(final_geo, Gravity.Center);
                        image.RePage();
                }
                collection.Write("filepath");
            }
            
The elapsed time that it took is pretty high as well 22542 ms.
By any chance, is it possible to reduce that somehow?
This time tends to result in timeout in my case since it's pretty slow like that and eventually, the purpose isn't served.

Before colection.Write("filepath") I tried using -
`var settings = new QuantizeSettings();
                settings.Colors = 256;
                collection.Quantize(settings);`
For 100 * 100, this, in turn, increased the memory consumption slightly and reduced the time drastically.
For 200 * 200, again the response time was considerably lowered but the size of the final image is bigger than the original one.
Original is 633 KB but the modified 200 * 200 is 740 KB.
                
This is the same behavior that I experienced with another gif (just 15 frames), 200 * 200 also causes an increase in size which is more than the original image.
               
![200](https://user-images.githubusercontent.com/32168989/136006738-5d1dbc09-b831-432c-8dbd-aa4b5f5725ed.gif)

**Comment by shivabhadauria** [comment]
Hi @dlemstra,

One small question, to fetch the frame count for gifs using Magick.NET, is there any built-in method or the only approach is -
```
var x = new MagickImageCollection(new FileInfo("FilePath"));
int frameCount = x.Count;
```
?

**Reply by dlemstra** [reply]
You could do this:

```C#
using (var images = new MagickImageCollection())
{
     images.Ping(FilePath);
     int frameCount = images.Count;
}
```
